src/isl/util.rs

```diff
@@ -0,0 +1,286 @@
+pub enum Range {
+    Decimal(RangeBoundaryValue, RangeBoundaryValue),
+    Float(RangeBoundaryValue, RangeBoundaryValue),
+    Integer(RangeBoundaryValue, RangeBoundaryValue),
+    IntegerNonNegative(RangeBoundaryValue, RangeBoundaryValue),
+    Timestamp(RangeBoundaryValue, RangeBoundaryValue),
+}
+
+impl Range {
+    pub fn decimal_range(
+        start: RangeBoundaryValue,
+        end: RangeBoundaryValue,
+    ) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::Decimal(start, end))
+    }
+
+    pub fn float_range(
+        start: RangeBoundaryValue,
+        end: RangeBoundaryValue,
+    ) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::Float(start, end))
+    }
+
+    pub fn int_range(start: RangeBoundaryValue, end: RangeBoundaryValue) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::Integer(start, end))
+    }
+
+    pub fn int_non_negative_range(
+        start: RangeBoundaryValue,
+        end: RangeBoundaryValue,
+    ) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::IntegerNonNegative(start, end))
+    }
+
+    pub fn timestamp_range(
+        start: RangeBoundaryValue,
+        end: RangeBoundaryValue,
+    ) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::Timestamp(start, end))
+    }
+
+    pub fn from_ion_element(value: &OwnedElement, is_non_negative: bool) -> IonSchemaResult<Range> {
+        let range = try_to!(value.as_sequence());
+        if range.len() != 2 {
+            return invalid_schema_error(
+                "Ranges must contain two values representing minimum and maximum ends of range.",
+            );
+        }
+
+        // set start of the range
+        let start = RangeBoundaryValue::from_ion_element(try_to!(range.get(0)), is_non_negative)?;
+
+        // set end of the range
+        let end = RangeBoundaryValue::from_ion_element(try_to!(range.get(1)), is_non_negative)?;
+
+        // validate both range boundary values
+        let range_type = Range::validate_range_boundaries(&start, &end)?;
+        match range_type {
+            RangeBoundaryValueType::Decimal(_) => Ok(Range::Decimal(start, end)),
+            RangeBoundaryValueType::Float(_) => Ok(Range::Float(start, end)),
+            RangeBoundaryValueType::Integer(_) => Ok(Range::Integer(start, end)),
+            RangeBoundaryValueType::IntegerNonNegative(_) => {
+                Ok(Range::IntegerNonNegative(start, end))
+            }
+            RangeBoundaryValueType::Timestamp(_) => Ok(Range::Timestamp(start, end)),
+        }
+    }
+
+    // helper method to which validates a non negative integer range boundary value
+    fn validate_non_negative_integer_range_boundary_value(
+        value: &AnyInt,
+    ) -> IonSchemaResult<usize> {
```

Comment:
The size of `usize` varies by platform and is usually used for addresses in memory or offsets. On some platforms, this would end up being a 4-byte integer.

Should we just return the same `AnyInt` so we can continue to support values of arbitrary size?

Reply:
If I understand this correctly, you are suggesting we should represent both `integer` and `non_negative_integers` as `AnyInt`. I think that should be okay as we already annotate range boundaries with it's type.